File: pkg_roscpp/src/simple_action_client_all.cpp
// refernce: https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1SimpleActionClient.html

#include "ros/ros.h"
#include "actionlib/client/simple_action_client.h"
#include "pkg_roscpp/CountNumbersAction.h"

class ActionClient{
protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionClient<pkg_roscpp::CountNumbersAction> ac_;

public:
    ActionClient() : 
            ac_("count_numbers", true)
    {
        ROS_INFO("waiting for server to start");
        ac_.waitForServer();    
    }

    ~ActionClient(void){}

    void send_goal(){
        ROS_INFO("connected to server");
        pkg_roscpp::CountNumbersGoal goal;
        goal.count_upto = 10;
        goal.wait_time = 1.0;
        ac_.sendGoal(goal, 
                boost::bind(&ActionClient::done_CB, this, _1, _2), 
                boost::bind(&ActionClient::active_CB, this), 
                boost::bind(&ActionClient::feedback_CB, this, _1));
        ROS_INFO("goal sent");
    }

    void done_CB(const actionlib::SimpleClientGoalState &state, 
            const pkg_roscpp::CountNumbersResultConstPtr &result)
    {
        //ROS_INFO("Finished in state: %s", (state).toString.c_str());
        ROS_INFO("Result: [%d]", (int)result->final_count);
    }

    void active_CB(){
        ROS_INFO("Goal just went active");        
    }

    void feedback_CB(const pkg_roscpp::CountNumbersFeedbackConstPtr &feedback){
        ROS_INFO("feedback received [%d]", feedback->current_count);
    }
};

int main(int argc, char **argv){
    ros::init(argc, argv, "simple_action_server");
    ActionClient Client;
    Client.send_goal();

    ros::spin();

    return 0;
}

File: pkg_rospy/scripts/simple_action_client_all.py
#! /usr/bin/env python

# Sends goal and does its work, done_cb callback function is invoked whenever a 
# msg on /count_numbers/result is received, feedback function is incoked 
# whenever a msg on /count_numbers/feedback is received

# reference: https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1simple__action__client_1_1SimpleActionClient.html

import rospy
import actionlib
from pkg_rospy.msg import CountNumbersAction, CountNumbersGoal

class CountNumbersClient():
    def __init__(self):
        self._ac = actionlib.SimpleActionClient('/count_numbers', CountNumbersAction)
        self._ac.wait_for_server()
        rospy.loginfo("connected to actionserver")

    def send_goal(self):
        goal = CountNumbersGoal()
        goal.count_upto = 10
        goal.wait_time = 1.0
        self._ac.send_goal(goal, done_cb = self.done_cb, feedback_cb=self.feedback_cb)
        rospy.loginfo("Goal has been sent")

        #rospy.sleep(1)     #uncomment to cancel goal while execution
        #self._ac.cancel_goal()
    
    def done_cb(self, status, result):
        rospy.loginfo("status is: " + str(status))
        rospy.loginfo("result is: " + str(result))  #3 means success, 4 means aborted, reference: http://docs.ros.org/en/kinetic/api/actionlib_msgs/html/msg/GoalStatus.html
        rospy.logwarn("Shutting down node")
        rospy.signal_shutdown("Done Executing")

    def feedback_cb(self, feedback):
        rospy.loginfo("feedback is: " + str(feedback))

def main():
    rospy.init_node("count_numbers_client")
    Client = CountNumbersClient()
    Client.send_goal()
    rospy.spin()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass